phoenix-core-client/src/main/java/org/apache/phoenix/compile/UpsertCompiler.java

```diff
@@ -842,11 +851,27 @@ public MutationPlan compile(UpsertStatement upsert) throws SQLException {
                         expression.getDataType(), column.getDataType(), "expression: "
                                 + expression.toString() + " in column " + column);
             }
+            if (!SchemaUtil.isPKColumn(column) && !isJsonNode(valueNode)) {
+                nonPKColumns.add(
+                        new Pair<>(ColumnName.caseSensitiveColumnName(column.getName().getString()),
+                                valueNode));
+            }
+            if (isJsonNode(valueNode)) {
+                jsonExpressions.add(
+                        new Pair<>(ColumnName.caseSensitiveColumnName(column.getName().getString()),
+                                valueNode));
+                jsonExpressions.addAll(nonPKColumns);
+                nonPKColumns = Lists.newArrayList();
```

Comment:
Why are you doing this here and resetting the nonPKColumns list ?

Reply:
You are adding again when you exit the loop. I didn't quite understand this logic.

Reply:
This is bascically say you have something like "UPSERT nonPkcol1,nonPkcol2,jsoncol1, jsoncol2,nonpkcol3,jsoncol3, nonpkcol4". as you find nonPK column add to list till you find a jsoncol at which point you include all those columns and reset because you want to add the next set of nonpk cols. and adding after exiting the loop to add the remaining nonpk cols which came after all the json cols.

Reply:
But ultimately everything merges to one list so why we need to reset. Why can't we just add all non pk cols in one list and json expressions in another list. At then end of the loop if json expressions list is not empty add all the non pk columns to json expression list.

Reply:
Yeah your comment perfectly makes sense